### PR TITLE
network: prevent interfaces to be initialized multiple times

### LINKED ISSUE #12315: DNS lookup broken in version 242 until restarting networkd
After booting, I have an IP and internet access (through a home router), but no DNS resolving. I use Arch Linux systemd-networkd from the testing repo (version 242), and I have systemd-resolved set up in the stub mode. The network conf is set to use DHCP. I have an ethernet device (enp5s0) and a bridge (br0) set up in a simple bridge configuration through systemd. Restarting systemd-networkd makes everything work fine. This seemingly started happening with 242 with no system configuration changes.

The previous version I was using is 241.93. Downgrading to that version seems to fix this issue, with some caveats: if I log into a non-graphical console after boot, and ping google right away, it fails to resolve. But if I ping again right after, it works. If I just log in to a DE, no apparent issues are visible for a while. However, at some point a DNS lookup will immediately fail (e.g., clicking a link while web browsing), but works upon trying again. This seems maybe related to issue #12142. I've been noticing this for a while, which is potentially the period of time I was using 241.7 (updated to on Feb 20).

When broken, output of resolvectl shows "Current Scopes: none" for the bridge, and no DNS servers. After networkd restart, scopes are "DNS LLMNR/IPv4 LLMNR/IPv6". 

Setting the networkd unit to use debug mode shows some seemingly relevant info and differences. This is the journalctl output during the initial boot: [systemd-networkd-broken.log](https://github.com/systemd/systemd/files/3077616/systemd-networkd-broken.log)

Note that enp5s0 gets renamed to eth0, then back to enp5s0 for some reason (not sure if this is normal). Afterward, there's a
`br0: Could not bring up interface: Invalid argument`
and
`enp5s0: Failed to add to bridge master's slave list: File exists`

And this is it when restarting it, which shows no such issues:
[systemd-networkd-good.log](https://github.com/systemd/systemd/files/3077617/systemd-networkd-good.log)

Here is resolved on boot (not working: 
[resolved-bad.log](https://github.com/systemd/systemd/files/3077619/resolved-bad.log)

And this is after restarting networkd: 
[resolved-good.log](https://github.com/systemd/systemd/files/3077620/resolved-good.log)

I'm using linux 5.0.7. Both r8169 and r8168 ethernet drivers give the same behavior. This did not start happening along with a kernel update.

edit:
Removing the bridge from the configuration fixes the issue too (in that it behaves as it did previously to 242). I also noticed that on occasional boots, even with the bridge, it behaves as in pre-242. Maybe it's a race condition?

### PR DESCRIPTION
**Pull request by yuwata**
When a uevent is received during the relevant interface is in `LINK_STATE_PENDING`, then the interface may be initialized twice. To prevent that, this introduces `LINK_STATE_INITIALIZED`.

Hopefully fixes #12315.

### CODE REVIEW

**Comment by poettering** [comment]
seems to break the centos CI tests.

File: src/network/networkd-link.h [hunk 1]
```diff
@@ -20,8 +20,9 @@
 #include "set.h"
 
 typedef enum LinkState {
-        LINK_STATE_PENDING,
-        LINK_STATE_CONFIGURING,
+        LINK_STATE_PENDING,     /* udev has not initialized the link. */
+        LINK_STATE_INITIALIZED, /* udev has initialized the link. */
+        LINK_STATE_CONFIGURING, /* configuring addresses, routes, etc. */
```
Comment (poettering) [inline src/network/networkd-link.h:25]: might be nice adding similar comments to the other states too...

**Comment by yuwata** [comment]
Updated.
- added one more commit that add logging link state change.
- added comments to remaining link states, requested by @poettering.
- hopefully fixed CI failure.

PTAL.

**Comment by yuwata** [comment]
In #12315, the reporter confirm this fixes the issue.
@poettering and @ssahani PTAL.

**Comment by davispuh** [comment]
This PR fixes it for me too :+1:

File: src/network/networkd-link.c [hunk 2]
```diff
@@ -609,8 +609,10 @@ static void link_detach_from_manager(Link *link) {
         if (!link || !link->manager)
                 return;
 
-        hashmap_remove(link->manager->links, INT_TO_PTR(link->ifindex));
-        set_remove(link->manager->links_requesting_uuid, link);
+        if (hashmap_get(link->manager->links, INT_TO_PTR(link->ifindex)) == link)
+                (void) hashmap_remove(link->manager->links, INT_TO_PTR(link->ifindex));
```
Comment (ssahani) [inline src/network/networkd-link.c]: Should we add the ```hashmap_get``` any way we not checking for ```hashmap_remove``` failure ?

File: src/network/networkd-link.c [hunk 3]
```diff
@@ -2531,6 +2515,42 @@ static void link_free_carrier_maps(Link *link) {
...
+
+        assert(link);
+
+        if (!link->manager || !netdev)
+                return;
+
+        if (link_get(link->manager, netdev->ifindex, &master) < 0)
+                return;
+
+        (void) hashmap_remove(master->slaves, INT_TO_PTR(link->ifindex));
```
Comment (ssahani) [inline src/network/networkd-link.c]: same here

**Comment by yuwata** [comment]
@ssahani Thank you for the comment. I've force-pushed a revised version. Now I hope the logic becomes safer. PTAL.

**Comment by yuwata** [comment]
I've updated whole logic of ref/unref `Link` object. PTAL.

**Comment by izenja** [comment]
Tested with the new commits and it still works.

File: src/network/networkd-link.h [hunk 4]
```diff
@@ -20,12 +20,13 @@
...
-        LINK_STATE_CONFIGURED,
-        LINK_STATE_UNMANAGED,
-        LINK_STATE_FAILED,
-        LINK_STATE_LINGER,
+        LINK_STATE_PENDING,     /* udev has not initialized the link. */
+        LINK_STATE_INITIALIZED, /* udev has initialized the link. */
+        LINK_STATE_CONFIGURING, /* configuring addresses, routes, etc. */
+        LINK_STATE_CONFIGURED,  /* everything is configured. */
+        LINK_STATE_UNMANAGED,   /* Unmanaged=yes is set. */
+        LINK_STATE_FAILED,      /* at least one configuration process fails. */
```
Comment (keszybz) [inline src/network/networkd-link.h]: fails → failed

File: src/network/networkd-link.h [hunk 5]
```diff
@@ -20,12 +20,13 @@
...
+        LINK_STATE_LINGER,      /* RTM_DELLINK for the link is received. */
```
Comment (keszybz) [inline src/network/networkd-link.h]: is received → has been received

**Comment by yuwata** [comment]
@keszybz Thanks!

**Pull request merged**